compiler/rustc_codegen_ssa/src/sir.rs

```diff
@@ -779,7 +802,9 @@ impl SirFuncCx<'tcx> {
             ty::Tuple(..) => self.lower_tuple_ty(bx, ty_layout),
             _ => ykpack::Ty::Unimplemented(format!("{:?}", ty_layout)),
         };
-        bx.cx().define_sir_type(sir_ty)
+        let tyid = bx.cx().define_sir_type(sir_ty);
+        self.layout_map.insert(tyid, (size, align));
```

Comment:
Does it make more sense to do this inside `define_sir_type` and store the layout map inside the `Sir` object of the codegen context?

Reply:
That way we don't duplicate work when using the same types in different `SirFuncCx`s

Reply:
I started doing it there first, but then realised putting it into `Sir` means the `layout_map` will be serialised, even though we have no need for it later on.

Reply:
I'm not sure that's right. Looks like we have fine-control over what gets serialised:
 * https://github.com/softdevteam/ykrustc/blob/6d1940069431afc538c031f94aafd1046bb9c36a/compiler/rustc_codegen_llvm/src/base.rs#L168-L169
 * https://github.com/softdevteam/ykrustc/blob/6d1940069431afc538c031f94aafd1046bb9c36a/compiler/rustc_codegen_llvm/src/sir.rs#L42-L54

Reply:
Ah, fair enough. I could put it into `Sir` then if you prefer, though I would have to rewrite most of this PR. Any strong argument why it's better there?

Reply:
That's unfortunate :\ sorry.

The reason it's better there is because each `SirFuncCx` is for one SIR `Body` and the same types can appear in multiple SIR bodies. Therefore if we compute the per-local layouts on a per-body basis, then we end up doing the same work many times.

`define_sir_type` (potentially a bad function name actually) only defines the type if we haven't seen that type before, otherwise it just returns the cached value. So you should be able to piggy back off that logic and only generate the layout if we haven't seen that type before?

Please check my reasoning though :)

Reply:
Sounds reasonable. Okay, I'll rewrite.

Reply:
We could do it in two stages if you like? We can merge this, and then the next PR optimises it?

Reply:
In fact, I think I'd prefer that in case what I proposed isn't possible for whatever reason.

Reply:
Are you okay with me changing `define_sir_type` then, as I'd have to pass the layout as arguments. Or do you want me to add another function to the `Sir`?

Reply:
Yes, that's fine if you need to.

We might be able to compute the layouts [here](https://github.com/softdevteam/ykrustc/blob/6d1940069431afc538c031f94aafd1046bb9c36a/compiler/rustc_codegen_ssa/src/sir.rs#L884). We can add your mapping to `SirTypes` without fear of it being serialized.

Only `SirTypes::map` is serialized.

Reply:
I don't think we can do it here, since we don't have access to the size and align values. Though `ykpack::Ty` has a `size` and `align` methods, we haven't implemented this for all Rust types, so for now we have to rely on the size and align we get from `ty_layout.layout.align/size`.